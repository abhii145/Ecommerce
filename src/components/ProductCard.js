import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { addToCart } from "../store/cartSlice";
import toast from "react-hot-toast";

const ProductCard = ({ product, }) => {
  const dispatch = useDispatch();

  const handleAdd = (product) => {
    toast.success("added to cart");
    dispatch(addToCart(product));
  };

  const { title, price, image } = product;
  const [isHovered, setIsHovered] = useState(false);
  return (
    <>
      <div
        className={`bg-white rounded-lg shadow-lg overflow-hidden m-4 sm:m-2 md:w-1/2 lg:w-1/3 xl:w-1/4 ${
          isHovered ? "transform scale-105" : ""
        } transition-transform duration-300 ease-in-out`}
        onMouseEnter={() => setIsHovered(true)}
        onMouseLeave={() => setIsHovered(false)}
      >
        <img
          src={image}
          alt={title}
          className="w-full h-48 object-contain object-center"
        />
        <div className="p-4">
          <h2 className="text-xl font-semibold mb-2">{title.slice(0, 52)}</h2>
          <p className="text-gray-600 text-sm">${price}</p>
          <button
            className="mt-2 bg-blue-500 text-white hover:bg-blue-600 py-2 px-4 rounded-full w-full"
            onClick={() => handleAdd(product)}
          >
            Add to Cart
          </button>
        </div>
        
      </div>
    </>
  );
};

export default ProductCard;
