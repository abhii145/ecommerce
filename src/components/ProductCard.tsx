import React, { memo } from "react"
import { FiInfo } from "react-icons/fi"
import { ProductCardProps } from "../types"
import { Link } from "react-router"
import { renderStars } from "../utils/renderStars"

const ProductCard = ({
  id,
  title,
  price,
  thumbnail,
  description,
  rating = 0,
}: ProductCardProps) => {
  return (
    <article className="bg-white rounded-lg shadow-md overflow-hidden w-full max-w-sm mx-auto transition-all duration-300 ease-in-out transform hover:shadow-lg hover:-translate-y-1">
      <div className="h-48 w-full bg-gray-100 flex items-center justify-center">
        <img
          src={thumbnail}
          alt={title}
          className="h-full object-contain"
          loading="lazy"
          width="100%"
          height="100%"
          data-testid="product-image"
        />
      </div>
      <div className="p-4 flex flex-col h-full">
        <h2
          className="text-xl font-semibold mb-2 truncate"
          data-testid="product-title"
        >
          {title}
        </h2>
        <p
          className="text-gray-600 mb-2 line-clamp-3 min-h-[4rem]"
          data-testid="product-description"
          aria-label={`Description: ${description}`}
        >
          {`${description?.slice(0, 50)}...`}
        </p>

        <div className="flex justify-between items-center mb-4">
          <span className="text-2xl font-bold" data-testid="product-price">
            â‚¹ {price.toFixed(2)}
          </span>
          <div
            className="flex items-center"
            data-testid="product-rating"
            aria-label={`Rating: ${rating} stars`}
          >
            {renderStars(rating)}
          </div>
        </div>

        <Link
          to={`/products/${id}`}
          className="w-full bg-gray-200 text-gray-700 px-4 py-2 rounded hover:bg-gray-300 transition-colors duration-300 flex items-center justify-center"
          aria-label={`More details about ${title}`}
          data-testid="more-details-link"
        >
          <FiInfo className="mr-2" />
          More Details
        </Link>
      </div>
    </article>
  )
}

export default memo(ProductCard)
