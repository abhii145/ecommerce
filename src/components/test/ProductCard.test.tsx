import React from "react"
import { render, screen } from "@testing-library/react"
import { BrowserRouter as Router } from "react-router"
import ProductCard from "../ProductCard"

const renderProductCard = (product) => {
  return render(
    <Router>
      <ProductCard {...product} />
    </Router>
  )
}

const commonProductProps = {
  id: 1,
  title: "Test Product",
  price: 100,
  thumbnail: "test.jpg",
  description:
    "This is a test product description that is quite long and should be truncated.",
  rating: 4.5,
}

describe("ProductCard Component", () => {
  it("renders ProductCard component", () => {
    renderProductCard(commonProductProps)

    expect(screen.getByTestId("product-title")).toBeInTheDocument()
    expect(screen.getByTestId("product-title")).toHaveTextContent(
      "Test Product"
    )

    expect(screen.getByTestId("product-price")).toBeInTheDocument()
    expect(screen.getByTestId("product-price")).toHaveTextContent("â‚¹ 100.00")

    expect(screen.getByTestId("product-description")).toBeInTheDocument()
    expect(screen.getByTestId("product-description")).toHaveTextContent(
      "This is a test product description that is quite l..."
    )

    const img = screen.getByTestId("product-image") as HTMLImageElement
    expect(img).toBeInTheDocument()
    expect(img.src).toContain("test.jpg")

    expect(screen.getByTestId("product-rating")).toBeInTheDocument()

    const moreDetailsLink = screen.getByTestId(
      "more-details-link"
    ) as HTMLAnchorElement
    expect(moreDetailsLink).toBeInTheDocument()
    expect(moreDetailsLink.href).toContain("/products/1")
  })

  it("renders ProductCard component with default rating", () => {
    const product = {
      ...commonProductProps,
      rating: undefined,
    }

    renderProductCard(product)
    expect(screen.getByTestId("product-rating")).toBeInTheDocument()
  })

  it("matches snapshot", () => {
    const { asFragment } = renderProductCard(commonProductProps)
    expect(asFragment()).toMatchSnapshot()
  })
})
