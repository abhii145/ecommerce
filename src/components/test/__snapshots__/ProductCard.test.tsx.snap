// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="bg-white rounded-lg shadow-md overflow-hidden w-full max-w-sm mx-auto transition-all duration-300 ease-in-out transform hover:shadow-lg hover:-translate-y-1"
  >
    <div
      class="h-48 w-full bg-gray-100 flex items-center justify-center"
    >
      <img
        alt="Test Product"
        class="h-full object-contain"
        data-testid="product-image"
        height="100%"
        loading="lazy"
        src="test.jpg"
        width="100%"
      />
    </div>
    <div
      class="p-4 flex flex-col h-full"
    >
      <h2
        class="text-xl font-semibold mb-2 truncate"
        data-testid="product-title"
      >
        Test Product
      </h2>
      <p
        aria-label="Description: This is a test product description that is quite long and should be truncated."
        class="text-gray-600 mb-2 line-clamp-3 min-h-[4rem]"
        data-testid="product-description"
      >
        This is a test product description that is quite l...
      </p>
      <div
        class="flex justify-between items-center mb-4"
      >
        <span
          class="text-2xl font-bold"
          data-testid="product-price"
        >
          â‚¹ 100.00
        </span>
        <div
          aria-label="Rating: 4.5 stars"
          class="flex items-center"
          data-testid="product-rating"
        >
          <div
            class="flex"
          >
            <svg
              class="text-yellow-500"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
            <svg
              class="text-yellow-500"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
            <svg
              class="text-yellow-500"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
            <svg
              class="text-yellow-500"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
            <svg
              class="text-yellow-500 opacity-50"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
          </div>
        </div>
      </div>
      <a
        aria-label="More details about Test Product"
        class="w-full bg-gray-200 text-gray-700 px-4 py-2 rounded hover:bg-gray-300 transition-colors duration-300 flex items-center justify-center"
        data-discover="true"
        data-testid="more-details-link"
        href="/products/1"
      >
        <svg
          class="mr-2"
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="12"
            x2="12"
            y1="16"
            y2="12"
          />
          <line
            x1="12"
            x2="12.01"
            y1="8"
            y2="8"
          />
        </svg>
        More Details
      </a>
    </div>
  </article>
</DocumentFragment>
`;
